### PR TITLE
vindexes: Efficient unicode hashing

### LINKED ISSUE #14398: Feature Request: Vendor the collation logic from `x/text` in Vitess
### Feature Description

One of the things we have discussed in the past, but we didn't track in an issue yet is vendoring the collation logic for `vindexes` from `x/text`. 

This logic is really old code in Vitess and today we would not write it in the way it exist, as we have the full encoding support that we can leverage now. We can't however switch to that, since we can't change `vindex` hashing as changing it would mean data ends up on a different shard. 

So the logic for `vindex` hash computation is / must be set in stone. That is also the issue here, since we already have been inadvertently "breaking" it in the past. With newer versions of Go and newer versions of `x/text`, the Unicode version has been changed. This means that certain character properties can be updated, which could lead to different behavior for say character normalization and denormalization. 

We have never seen that reported as an issue though, which indicates that text based case insensitive columns used in vindexes are usually often ASCII only (like a token), or using characters with say accents from the lower plains where changes hardly ever happen in new Unicode releases. 

To avoid this problem in the future going forward, we need to vendor this logic. Vendoring will also allow for large and significant performance optimizations since we can remove intermediate steps from the hashing process. 

### Use Case(s)

Any usage of `unicode_loose_*` vindexes.

### PR DESCRIPTION
**Pull request by vmg**
## Description

The current implementation of Unicode vindexes is not efficient because it depends on the `x/text/collate` package, which allocates a linear amount of memory based on the length of the input string being collated. This can be done much more efficiently because the value we require for a vindex is always a hash, so it can be calculated using a constant amount of memory if instead of generating a weight string and hashing it, we write directly into a hasher.

Paired on this with @dbussink.

Benchmarks:

```
goos: linux
goarch: amd64
pkg: vitess.io/vitess/go/vt/vtgate/vindexes
cpu: 13th Gen Intel(R) Core(TM) i9-13900K
                                │   old.txt    │               new.txt               │
                                │    sec/op    │   sec/op     vs base                │
UnicodeTest/repeat=1/MD5-16       10.003µ ± 2%   7.619µ ± 0%  -23.83% (p=0.000 n=10)
UnicodeTest/repeat=1/xxHash-16     8.714µ ± 3%   6.547µ ± 1%  -24.86% (p=0.000 n=10)
UnicodeTest/repeat=4/MD5-16        39.70µ ± 2%   29.96µ ± 1%  -24.55% (p=0.000 n=10)
UnicodeTest/repeat=4/xxHash-16     34.93µ ± 2%   26.13µ ± 1%  -25.21% (p=0.000 n=10)
UnicodeTest/repeat=16/MD5-16       159.7µ ± 3%   119.8µ ± 1%  -24.97% (p=0.000 n=10)
UnicodeTest/repeat=16/xxHash-16    135.8µ ± 4%   102.7µ ± 1%  -24.32% (p=0.000 n=10)
geomean                            37.13µ        27.99µ       -24.62%

                                │     old.txt     │              new.txt               │
                                │      B/op       │    B/op     vs base                │
UnicodeTest/repeat=1/MD5-16        12906.00 ±  7%   80.00 ± 0%  -99.38% (p=0.000 n=10)
UnicodeTest/repeat=1/xxHash-16     12814.50 ±  5%   72.00 ± 0%  -99.44% (p=0.000 n=10)
UnicodeTest/repeat=4/MD5-16        49138.50 ±  5%   80.00 ± 0%  -99.84% (p=0.000 n=10)
UnicodeTest/repeat=4/xxHash-16     50670.00 ±  4%   72.00 ± 0%  -99.86% (p=0.000 n=10)
UnicodeTest/repeat=16/MD5-16      194932.50 ± 11%   80.00 ± 0%  -99.96% (p=0.000 n=10)
UnicodeTest/repeat=16/xxHash-16   195962.50 ± 11%   72.00 ± 0%  -99.96% (p=0.000 n=10)
geomean                             48.88Ki         75.89       -99.85%

                                │   old.txt    │              new.txt               │
                                │  allocs/op   │ allocs/op   vs base                │
UnicodeTest/repeat=1/MD5-16        34.000 ± 0%   2.000 ± 0%  -94.12% (p=0.000 n=10)
UnicodeTest/repeat=1/xxHash-16     34.000 ± 0%   2.000 ± 0%  -94.12% (p=0.000 n=10)
UnicodeTest/repeat=4/MD5-16       133.000 ± 0%   2.000 ± 0%  -98.50% (p=0.000 n=10)
UnicodeTest/repeat=4/xxHash-16    133.000 ± 0%   2.000 ± 0%  -98.50% (p=0.000 n=10)
UnicodeTest/repeat=16/MD5-16      529.000 ± 0%   2.000 ± 0%  -99.62% (p=0.000 n=10)
UnicodeTest/repeat=16/xxHash-16   529.000 ± 0%   2.000 ± 0%  -99.62% (p=0.000 n=10)
geomean                             133.7        2.000       -98.50%
```

cc @mattlord @aquarapid 

## Related Issue(s)

Fixes #14398

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on the CI
-   [ ] Documentation was added or is not required

### CODE REVIEW

**Comment by dbussink** [comment]
> This can be done much more efficiently because the value we require for a vindex is always a hash, so it can be calculated using a constant amount of memory if instead of generating a weight string and hashing it, we write directly into a hasher.

The vendoring also addresses another issue. When `x/text` is updated, the collation tables can be fixed for a newer Unicode version but that could (very subtlety) break a vindex since it changes what input hashes to.

Here we vendor, so we essentially set it in stone and keep it always consistent for the future.

File: go/mysql/collations/vindex/collate/golden_test.go [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package collate
+
+import "testing"
+
+func TestGenerateGolden(t *testing.T) {
+
+}
```
Comment (harshit-gangal) [inline go/mysql/collations/vindex/collate/golden_test.go:7]: am I missing something here?

**Review by dbussink — APPROVED** [review]
So much more memory efficient!

File: go/mysql/collations/vindex/collate/golden_test.go (hunk 1 above)
Comment (dbussink) [reply go/mysql/collations/vindex/collate/golden_test.go:7]: Looks like this shouldn't be there.

**Review by harshit-gangal — COMMENTED** [review]
A general question. Does the output for the current hash functions still remain the same?

**Review by harshit-gangal — APPROVED** [review]

**Comment by vmg** [comment]
Yes, this is 100% backwards compatible and there are golden tests that verify it. Otherwise we'd break all existing Vitess deployments by changing the shard position of existing values!

**Pull request merged**